### PR TITLE
20210225 ontag cmdlet

### PR DESCRIPTION
**Pull request by changlong-liu**

### CODE REVIEW

File: src/generate/renders/tests/CliTestCmdlet.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                                name: functionName,
+                                lines: commandLines
+                                    .slice(0, -1)
+                                    .map(
+                                        (x) => x.split('{').join('{{').split('}').join('}}') + ' ',
+                                    ),
+                                lastLine: commandLines.last,
+                            });
+                        }
+        return ret;
```
Comment (qiaozha) [inline src/generate/renders/tests/CliTestCmdlet.ts:97]: @changlong-liu Could you add some unit tests for this. Such as the get data logic and the render logic? You may refer to commands.py related test under unittest folder for more details.

File: src/generate/generators/ExtensionIncre.ts [hunk 2]
```diff
@@ -145,5 +148,30 @@ export class AzExtensionIncrementalGenerator extends GeneratorBase {
...
             );
         GenerateMetaFile(this.model);
+        await this.generateIncrementalSingleAndAddtoOutput(
+            new CliCmdletTest(this.model, false),
+            true,
+        );
+        await this.generateIncrementalSingleAndAddtoOutput(
+            new CliCmdletTest(this.model, true),
+            true,
+        );
```
Comment (qiaozha) [inline src/generate/generators/ExtensionIncre.ts:158]: how about
```
for(const boolVal of [false, true]) {
      await this.generateIncrementalSingleAndAddtoOutput(
            new CliCmdletTest(this.model, boolVal),
            true,
        );
} 
```

File: src/generate/CodeModelAzImpl.ts [hunk 3]
```diff
@@ -2973,6 +2976,15 @@ export class CodeModelCliImpl implements CodeModelAz {
...
                 // }
+                else if (this.extensionName == 'vm') {
+                    if (paramName == 'vm_name') paramName = 'name';
+                } else if (this.extensionName == 'network') {
+                    if (paramName == 'network_security_group_name') paramName = 'nsg_name';
+                    else if (paramName == 'virtual_network_name') paramName = 'vnet_name';
+                    else if (commandGroup.toLowerCase() === 'privatelinkservices') {
+                        if (paramName == 'service_name') paramName = 'name';
+                    }
+                }
```
Comment (qiaozha) [inline src/generate/CodeModelAzImpl.ts:2987]: Wondering why do we have this part? is this because of cli existing command name and swagger name mapping? or something else? What if we found more mismatch? Should we put it into a configurable file ?

File: src/generate/generators/CoreFull.ts [hunk 4]
```diff
@@ -120,6 +124,33 @@ export class AzCoreFullGenerator extends GeneratorBase {
...
+                            PathConstants.conftestFile,
+                        ),
+                        path.join(
+                            PathConstants.templateRootFolder,
+                            PathConstants.testFolder,
+                            PathConstants.cmdletFolder,
+                            PathConstants.conftestFile + '.njx',
+                        ),
+                    ),
+                );
```
Comment (qiaozha) [inline src/generate/generators/CoreFull.ts:153]: I think the path join part can be moved into the template definition instead of each generators.

File: src/generate/renders/TemplateBase.ts [hunk 5]
```diff
@@ -33,3 +33,24 @@ export abstract class TemplateBase {
 
     public abstract async GetRenderData(model: CodeModelAz): Promise<any>;
 }
+
+export class SimpleTemplate extends TemplateBase {
```
Comment (qiaozha) [inline src/generate/renders/TemplateBase.ts:37]: Just curious why don't we put it as a single file ?

File: src/generate/renders/tests/CliTestStep.ts [hunk 6]
```diff
@@ -218,6 +217,9 @@ export class CliTestStep extends TemplateBase {
     }
 
     public static parameterLine(parameterNames, withChecksDef = false) {
+        if (!AzConfiguration.getValue(CodeGenConstants.az)[CodeGenConstants.useTestStepParam]) {
```
Comment (qiaozha) [inline src/generate/renders/tests/CliTestStep.ts:220]: Can we use === true to decide whether user has set the correct value ? just in case they put a lot of strange values

File: test/scenarios/attestation/output/src/attestation/azext_attestation/tests/cmdlet/conftest.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+def pytest_runtest_logreport(report):
+    report.nodeid = "..." + report.nodeid[report.nodeid.rfind("::"):]
```
Comment (qiaozha) [inline test/scenarios/attestation/output/src/attestation/azext_attestation/tests/cmdlet/conftest.py:2]: wondering who will invoke this file ?

**Comment by qiaozha** [comment]
@changlong-liu Left a few comment, I am wondering should we make positive and negative test configurable?

File: src/generate/CodeModelAzImpl.ts (hunk 3 above)
Comment (changlong-liu) [reply src/generate/CodeModelAzImpl.ts:2987]: this is a quick fix for findings in depth coverage. they are existing param hard rule in CLI. I think they are similar with the mapping resource_group_name --> resource_group.

 I think cli-directive is a solution to configure this. But the problem will become who should configure these hard rules.

File: src/generate/generators/CoreFull.ts (hunk 4 above)
Comment (changlong-liu) [reply src/generate/generators/CoreFull.ts:153]: Here SimpleTemplate don't bind to a particular template or output path. 
This kind of render is designed for those files which don't have any dynamic data, so that we don't need write a tiny render for several templates for many times. For example test/__init__.py files and this pytest configuration file.

Hi @qiaozha , @lirenhe ,
And I think the actual pain here is that we need to write generation code of one file for four times in CoreFull, CoreIncre, ExtensionFull, ExtensionIncre... I think it will be better to have some envelop in GeneratorBase. 
I can refactor about this if you think it's the right direction.

File: src/generate/renders/TemplateBase.ts (hunk 5 above)
Comment (changlong-liu) [reply src/generate/renders/TemplateBase.ts:37]: I'm neutral on whether need to put it in a single file.
I put it in this file since I think this is an common render which may be used for many .njx

File: test/scenarios/attestation/output/src/attestation/azext_attestation/tests/cmdlet/conftest.py (hunk 7 above)
Comment (changlong-liu) [reply test/scenarios/attestation/output/src/attestation/azext_attestation/tests/cmdlet/conftest.py:2]: This file is used to set pytest output format

**Comment by changlong-liu** [comment]
> @changlong-liu Changlong Liu FTE Left a few comment, I am wondering should we make positive and negative test configurable?

yes, we can discuss in CLI bi-weekly meeting whether mock test should be generate by default or on demand.

File: test/unittest/render-commands.py.ts [hunk 8]
```diff
@@ -316,7 +316,7 @@ describe('renderCommandsPYTest', () => {
         });
         const oriFile = path.join(
             `${__dirname}`,
-            '../../test/unittest/expected/generated/ori_commands.py',
```
Comment (qiaozha) [inline test/unittest/render-commands.py.ts:319]: why do we change this?

File: test/unittest/render-cmdlets.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        result = await readFile(oriFile);
+        const expectedFile = path.join(
+            `${__dirname}`,
+            '../../test/unittest/expected/tests/cmdlet/test_negative.py',
+        );
+        const expected = await readFile(expectedFile);
+        assert.deepStrictEqual(result, expected, 'render logic in test_negative.py is incorrect');
+        await rmFile(oriFile);
+    });
+});
```
Comment (qiaozha) [inline test/unittest/render-cmdlets.ts:122]: Could we add some corner case such as, when we get empty data as no swagger example was provided or multiple operations point to the same example or multi examples refered by the same operation.

File: test/scenarios/attestation/configuration/readme.az.md [hunk 10]
```diff
@@ -7,8 +7,10 @@ az:
   extensions: attestation
   package-name: azure-mgmt-attestation
   namespace: azure.mgmt.attestation
+  use-test-step-param: true
 az-output-folder: $(azure-cli-extension-folder)/src/attestation
 python-sdk-output-folder: "$(az-output-folder)/azext_attestation/vendored_sdks/attestation"
+gen-cmdlet-test: true
```
Comment (qiaozha) [inline test/scenarios/attestation/configuration/readme.az.md:13]: Could we use a relatively complex RP for this scenario test? just don't think this is a good candidate to test some corner case such as polymorphism or complex resource dependencies

File: src/utils/helper.ts [hunk 11]
```diff
@@ -792,3 +792,9 @@ export function checkNested(obj, path: string) {
     }
     return true;
 }
+
+export function isTrue(t: any) {
+    return (
+        !isNullOrUndefined(t) && (t.toString().toLowerCase() === 'true' || typeof t === 'object')
```
Comment (qiaozha) [inline src/utils/helper.ts:798]: Does the function name IsTrue seem a little inappropriate for where the condition could be true when it's a object ?

File: src/generate/CodeModelAzImpl.ts (hunk 3 above)
Comment (qiaozha) [reply src/generate/CodeModelAzImpl.ts]: I don't see the logic of handling this mapping part now?

File: src/generate/generators/CoreFull.ts (hunk 4 above)
Comment (qiaozha) [reply src/generate/generators/CoreFull.ts]: I think we shuuld log a issue for that. and fix it in other PR ?

**Comment by qiaozha** [comment]
@changlong-liu I am also curious, Since you are using code template to generate the conftest.py or positive negative tests, Is this possible that the generated code may violate some pylint rules? Do we need to use code format tools to format them just in case ?

File: src/generate/CodeModelAzImpl.ts (hunk 3 above)
Comment (changlong-liu) [reply src/generate/CodeModelAzImpl.ts]: yes, handle this in RP's readme.az.md with cli-directive

File: src/generate/generators/CoreFull.ts (hunk 4 above)
Comment (changlong-liu) [reply src/generate/generators/CoreFull.ts]: sure https://github.com/Azure/autorest.az/issues/776

File: test/unittest/render-commands.py.ts (hunk 8 above)
Comment (changlong-liu) [reply test/unittest/render-commands.py.ts:319]: it use a same filename with other testcase, causing issue occasionally

**Comment by changlong-liu** [comment]
> @changlong-liu Changlong Liu FTE I am also curious, Since you are using code template to generate the conftest.py or positive negative tests, Is this possible that the generated code may violate some pylint rules? Do we need to use code format tools to format them just in case ?

Hi @qiaozha , don't get pylint issue now. And I didn't use auto-formating on them since:
1. it will affect inplace-edit in test file
2. the formatted result looks not as good as current output. (now each param is in one txt line, and the auto-formatting will merge them without particular alignment)

File: test/unittest/render-cmdlets.ts (hunk 9 above)
Comment (changlong-liu) [reply test/unittest/render-cmdlets.ts:135]: Added an empty case, seem multi operations point to same example and multi examples referred by same operation handled by modelerfour/codemodel so don't have relation with this unit

File: test/scenarios/attestation/configuration/readme.az.md (hunk 10 above)
Comment (changlong-liu) [reply test/scenarios/attestation/configuration/readme.az.md:13]: added datafactory

File: test/unittest/render-commands.py.ts (hunk 8 above)
Comment (qiaozha) [reply test/unittest/render-commands.py.ts:319]: I see

**Comment by qiaozha** [comment]
> > @changlong-liu Changlong Liu FTE I am also curious, Since you are using code template to generate the conftest.py or positive negative tests, Is this possible that the generated code may violate some pylint rules? Do we need to use code format tools to format them just in case ?
> 
> Hi @qiaozha , don't get pylint issue now. And I didn't use auto-formating on them since:
> 
> 1. it will affect inplace-edit in test file
> 2. the formatted result looks not as good as current output. (now each param is in one txt line, and the auto-formatting will merge them without particular alignment)

I see

**Review by qiaozha — APPROVED** [review]
